kroxylicious-filters/kroxylicious-record-validation/src/test/java/io/kroxylicious/proxy/filter/validation/RecordValidationFilterTest.java

```diff
@@ -0,0 +1,170 @@
+
+    @Mock
+    private ProduceRequestValidator produceRequestValidator;
+
+    @Mock
+    private TopicValidationResult topicValidationResult;
+
+    @Mock(strictness = LENIENT)
+    private FilterContext context;
+
+    @Captor
+    private ArgumentCaptor<ApiMessage> apiMessageCaptor;
+
+    RecordValidationFilterTest() {
+    }
+
+    @BeforeEach
+    void setUp() {
+        when(context.forwardRequest(any(RequestHeaderData.class), apiMessageCaptor.capture())).then(invocationOnMock -> {
+            var filterResult = mock(RequestFilterResult.class);
+            lenient().when(filterResult.message()).thenReturn(apiMessageCaptor.getValue());
+            return CompletableFuture.completedFuture(filterResult);
+        });
+
+        when(context.requestFilterResultBuilder()).then(invocationOnMock -> {
+            var builder = mock(RequestFilterResultBuilder.class);
+            var filterResult = mock(RequestFilterResult.class);
+
+            var closeOrTerminalStage = mock(CloseOrTerminalStage.class);
+            lenient().when(closeOrTerminalStage.completed()).thenReturn(CompletableFuture.completedStage(filterResult));
+            lenient().when(closeOrTerminalStage.build()).thenReturn(filterResult);
+
+            when(builder.shortCircuitResponse(apiMessageCaptor.capture())).then(invocation -> {
+                lenient().when(filterResult.shortCircuitResponse()).thenReturn(true);
+                lenient().when(filterResult.message()).thenReturn(apiMessageCaptor.getValue());
+                return closeOrTerminalStage;
+            });
+            return builder;
+        });
+    }
+
+    @Test
+    void rejectsNullValidator() {
+        assertThatThrownBy(() -> new RecordValidationFilter(false, null))
+                .isInstanceOf(IllegalArgumentException.class);
+    }
+
+    @Test
+    void requestThatPassesValidationIsForwarded() {
+        // Given
+        var validator = new RecordValidationFilter(false, produceRequestValidator);
+
+        when(topicValidationResult.isAnyPartitionInvalid()).thenReturn(false);
+
+        var header = new RequestHeaderData();
+        var request = buildProduceRequestData(new ProduceRequestData.TopicProduceData()
+                .setName(MY_TOPIC)
+                .setPartitionData(List.of(new ProduceRequestData.PartitionProduceData())));
+
+        when(produceRequestValidator.validateRequest(request)).thenReturn(
+                CompletableFuture.completedStage(new ProduceRequestValidationResult(Map.of(MY_TOPIC, topicValidationResult))));
+
+        // When
+        var result = validator.onProduceRequest(HIGHEST_SUPPORTED_VERSION, header, request, context);
+
+        // Then
+        assertThat(result)
+                .succeedsWithin(Duration.ofSeconds(1))
+                .satisfies(rfr -> {
+                    assertThat(rfr.shortCircuitResponse()).isFalse();
+                    assertThat(rfr.message()).isEqualTo(request);
+                });
```

Comment:
Certainly not for this PR but I do wonder if we would get value from spending some time building some Kroxylicious specific conditions/assertions (see https://assertj.github.io/doc/#assertj-core-extensions)

Reply:
Yeah.. the amount of effort requires to mock FilterContext is unreasonable.  I wonder whether a stub implementation or at least a pre-wired mock would help.

It turns out to be tricky to unit test RVF.  I think I'd like to come back to it with fresh eyes.  Maybe be the abstractions could be improved so they are more amenable to unit testing.